internal/pkg/gateway/api.go

```diff
@@ -11,64 +11,78 @@ import (
 	"fmt"
 	"sync"
 
+	"github.com/golang/protobuf/proto"
 	gp "github.com/hyperledger/fabric-protos-go/gateway"
 	"github.com/hyperledger/fabric-protos-go/peer"
 	"github.com/hyperledger/fabric/protoutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
 type endorserResponse struct {
 	pr  *peer.ProposalResponse
-	err error
+	err *gp.EndpointError
 }
 
 // Evaluate will invoke the transaction function as specified in the SignedProposal
 func (gs *Server) Evaluate(ctx context.Context, proposedTransaction *gp.ProposedTransaction) (*gp.Result, error) {
 	if proposedTransaction == nil {
-		return nil, fmt.Errorf("a proposed transaction is required")
+		return nil, status.Error(codes.InvalidArgument, "a proposed transaction is required")
 	}
 	signedProposal := proposedTransaction.Proposal
 	channel, chaincodeID, err := getChannelAndChaincodeFromSignedProposal(proposedTransaction.Proposal)
 	if err != nil {
-		// TODO need to specify status codes
-		return nil, fmt.Errorf("failed to unpack channel header: %w", err)
+		return nil, status.Errorf(codes.InvalidArgument, "failed to unpack transaction proposal: %s", err)
 	}
 
 	endorsers, err := gs.registry.endorsers(channel, chaincodeID)
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Unavailable, "%s", err)
 	}
 	if len(endorsers) == 0 {
-		return nil, fmt.Errorf("no endorsing peers found for channel: %s", proposedTransaction.ChannelId)
+		return nil, status.Errorf(codes.Unavailable, "no endorsing peers found for channel: %s", proposedTransaction.ChannelId)
```

Comment:
`Unavailable` or `NotFound`? What you have may be the most appropriate but I'm not sure given the text of the message. This applies to the other blocks that do the same thing below.

Reply:
Changed to NotFound